tests/rest/client/test_auth.py

```diff
@@ -669,6 +669,53 @@ def test_refresh_token_expiry(self):
         refresh_response = self.use_refresh_token(refresh_token)
         self.assertEqual(refresh_response.code, 403, refresh_response.result)
 
+    @override_config(
+        {
+            "access_token_lifetime": "2m",
+            "refresh_token_lifetime": "2m",
+            "session_lifetime": "3m",
+        }
+    )
+    def test_ultimate_session_expiry(self):
+        """
+        The session can be configured to have an ultimate, limited lifetime.
+        """
+
+        body = {"type": "m.login.password", "user": "test", "password": self.user_pass}
+        login_response = self.make_request(
+            "POST",
+            "/_matrix/client/r0/login?org.matrix.msc2918.refresh_token=true",
+            body,
+        )
+        self.assertEqual(login_response.code, 200, login_response.result)
+        refresh_token = login_response.json_body["refresh_token"]
+
+        # Advance shy of 2 minutes into the future
+        self.reactor.advance(119.0)
+
+        # Refresh our access token. It should still be valid right now.
+        refresh_response = self.use_refresh_token(refresh_token)
+        self.assertEqual(refresh_response.code, 200, refresh_response.result)
+        self.assertIn(
+            "refresh_token",
+            refresh_response.json_body,
+            "No new refresh token returned after refresh.",
+        )
+        # Notice that our access token lifetime has been diminished to match the
+        # session length.
+        # 3 minutes - 119 seconds = 61 seconds.
+        self.assertEqual(refresh_response.json_body["expires_in_ms"], 61_000)
```

Comment:
Didn't realise underscores were old enough to be in 3.6!

Reply:
I still remember when underscores and f-strings were too new and I couldn't use them in my 6th form college project because the computers at the 6th form college didn't have a new enough Python ... now look where we are :)